homeassistant/components/teslemetry/sensor.py

```diff
@@ -448,66 +523,153 @@ async def async_setup_entry(
-                TeslemetryEnergyHistorySensorEntity(energysite, description)
-                for energysite in entry.runtime_data.energysites
-                for description in ENERGY_HISTORY_DESCRIPTIONS
-                if energysite.history_coordinator
-            ),
-        )
+    entities: list[SensorEntity] = []
+    for vehicle in entry.runtime_data.vehicles:
+        for description in VEHICLE_DESCRIPTIONS:
+            if (
+                not vehicle.api.pre2021
+                and description.streaming_key
+                and vehicle.firmware >= description.streaming_firmware
+            ):
+                entities.append(TeslemetryStreamSensorEntity(vehicle, description))
+            elif description.polling:
+                entities.append(TeslemetryVehicleSensorEntity(vehicle, description))
+
+        for time_description in VEHICLE_TIME_DESCRIPTIONS:
+            if (
+                not vehicle.api.pre2021
+                and vehicle.firmware >= time_description.streaming_firmware
+            ):
+                entities.append(
+                    TeslemetryStreamTimeSensorEntity(vehicle, time_description)
+                )
+            else:
+                entities.append(
+                    TeslemetryVehicleTimeSensorEntity(vehicle, time_description)
+                )
+
+    entities.extend(
+        TeslemetryEnergyLiveSensorEntity(energysite, description)
+        for energysite in entry.runtime_data.energysites
+        for description in ENERGY_LIVE_DESCRIPTIONS
+        if description.key in energysite.live_coordinator.data
+    )
+
+    entities.extend(
+        TeslemetryWallConnectorSensorEntity(energysite, din, description)
+        for energysite in entry.runtime_data.energysites
+        for din in energysite.live_coordinator.data.get("wall_connectors", {})
+        for description in WALL_CONNECTOR_DESCRIPTIONS
     )
 
+    entities.extend(
+        TeslemetryEnergyInfoSensorEntity(energysite, description)
+        for energysite in entry.runtime_data.energysites
+        for description in ENERGY_INFO_DESCRIPTIONS
+        if description.key in energysite.info_coordinator.data
+    )
+
+    entities.extend(
+        TeslemetryEnergyHistorySensorEntity(energysite, description)
+        for energysite in entry.runtime_data.energysites
+        for description in ENERGY_HISTORY_DESCRIPTIONS
+        if energysite.history_coordinator is not None
+    )
+
+    async_add_entities(entities)
+
+
+class TeslemetryStreamSensorEntity(TeslemetryVehicleStreamEntity, RestoreSensor):
+    """Base class for Teslemetry vehicle streaming sensors."""
+
+    entity_description: TeslemetryVehicleSensorEntityDescription
+
+    def __init__(
+        self,
+        data: TeslemetryVehicleData,
+        description: TeslemetryVehicleSensorEntityDescription,
+    ) -> None:
+        """Initialize the sensor."""
+        self.entity_description = description
+        assert description.streaming_key
```

Comment:
If we assert it to be there, why is it optional in the first place?

Reply:
Some entities are streaming, some are polling, some are both. The `async_setup_entry` makes a decision about which class type to create for the description based on what is supported. This means a description may not have a streaming key, however async_setup_entry will never try to create the class without it.